gt-scheduler/website
# Send Email Invitation to Friends

### Summary

Resolves #163 , Firebase Function PR: https://github.com/gt-scheduler/firebase-conf/pull/2


- Added a new firebase collection (`friend-invites`) that keeps track of pending schedule invites (information about senderId, friendId, term, version).
- Created a firebase cloud function that checks arguments, creates a new invite record, and sends an email with a backlink generated from the new record's id (`gt-scheduler.org/invite/<record_id>`).
- Setup nodemailer to send emails
- Added react-dom-node to handle dynamic routes to support the invite backlink
- Created a firebase cloud function that deletes the invite record when the backlink is clicked. It is called by the dynamic route. This cloud function or the function in the dynamic route can be extended to perform the actual schedule-sharing updates

### Checklist
- [x] An email is sent to the right user. If that user does not exist in the database, return an error.
- [x] The email must have a message (such as, who sent them the invitation? to view which schedule? of which semester? etc.)
- [x] The email should contain a click-back link (Note: we will implement what this click-back link would do later but right now just know that we'll use this email to confirm/establish the "friendship")

### New Cloud Functions
New Cloud Functions expect the following in the request body:
- create_friend_invitation:
```
{
    IDToken: string,        // Firebase auth token of current user ( await auth.currentUser?.getIdToken() ),
    friendEmail: string,
    term: string,           // can be found in ScheduleContext
    version: string,        // schedule version id; can be found in ScheduleContext
}
```

- handle_friend_invitation:
```
{
    inviteId: string        // id of valid record in friend-invites collections
}
```

### How to Test
- Follow the steps in the `gt-scheduler/firebase-conf` repo to run the firebase emulator
- Sign in with email (have at least 2 accs)
- Use the input box at the bottom of the Course Search tab to enter the second email and then press the button. Check the emulator logs and db updates
- Check your email for the backlink and press it. Monitor the logs and db updates again
